src/librbd/image/CreateRequest.cc

```diff
@@ -0,0 +1,678 @@
+
+  if ((*result < 0) && (*result != -ENOENT)) {
+    lderr(m_cct) << "failed to retrieve mirror mode: " << cpp_strerror(*result) << dendl;
+
+    m_r_saved = *result;
+    CreateRequest<I>::remove_object_map();
+    return nullptr;
+  }
+
+  cls::rbd::MirrorMode mirror_mode_internal = cls::rbd::MIRROR_MODE_DISABLED;
+  if (*result == 0) {
+    bufferlist::iterator it = m_outbl.begin();
+    *result = cls_client::mirror_mode_get_finish(&it, &mirror_mode_internal);
+    if (*result < 0) {
+      lderr(m_cct) << "Failed to retrieve mirror mode" << dendl;
+
+      m_r_saved = *result;
+      CreateRequest<I>::remove_object_map();
+      return nullptr;
+    }
+  }
+
+  // TODO: remove redundant code...
+  switch (mirror_mode_internal) {
+  case cls::rbd::MIRROR_MODE_DISABLED:
+  case cls::rbd::MIRROR_MODE_IMAGE:
+  case cls::rbd::MIRROR_MODE_POOL:
+    m_mirror_mode = static_cast<rbd_mirror_mode_t>(mirror_mode_internal);
+    break;
+  default:
+    lderr(m_cct) << "Unknown mirror mode ("
+               << static_cast<uint32_t>(mirror_mode_internal) << ")"
+               << dendl;
+    *result = -EINVAL;
+    CreateRequest<I>::remove_object_map();
+    return nullptr;
+  }
+
+  journal_create();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::journal_create() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  Context *ctx = new FunctionContext([this](int r) {
+      r = Journal<I>::create(m_image_ctx->md_ctx, m_image_ctx->id,
+                             m_journal_order, m_journal_splay_width, m_journal_pool,
+                             m_force_non_primary, m_primary_mirror_uuid);
+      CreateRequest<I>::handle_journal_create(r);
+    });
+
+  m_image_ctx->op_work_queue->queue(ctx, 0);
+}
+
+template<typename I>
+void CreateRequest<I>::handle_journal_create(int r) {
+  ldout(m_cct, 20) << __func__ << ": r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "error creating journal: " << cpp_strerror(r) << dendl;
+
+    m_r_saved = r;
+    CreateRequest<I>::remove_object_map();
+    return;
+  }
+
+  fetch_mirror_image();
+}
+
+template<typename I>
+void CreateRequest<I>::fetch_mirror_image() {
+  if ((m_mirror_mode != RBD_MIRROR_MODE_POOL) && !m_force_non_primary) {
+    complete(0);
+    return;
+  }
+
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_image_get_start(&op, m_image_ctx->id);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_fetch_mirror_image>(this);
+  m_outbl.clear();
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_outbl);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_fetch_mirror_image(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  // this should always be the case during create
+  assert(*result == -ENOENT);
```

Comment:
Not a safe assumption -- better to fail gracefully when dealing with out-of-band data inputs.